src/compiler/checker.ts

```diff
@@ -20352,16 +20353,16 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function getBestMatchIndexedAccessTypeOrUndefined(source: Type, target: Type, nameType: Type) {
-        const idx = getIndexedAccessTypeOrUndefined(target, nameType);
-        if (idx) {
-            return idx;
-        }
         if (target.flags & TypeFlags.Union) {

```

Comment:
swapping the order of operations here improves some errors quite a bit in my eyes since the union discrimination happens early, for example:
```ts
const obj6: {
  prop:
    | {
        type: "foo";
        prop: string;
      }
    | {
        type: "bar";
        prop: number;
      };
} = {
  prop: {
    type: "foo",
    // current: Type 'boolean' is not assignable to type 'string | number'.
    // pr: Type 'boolean' is not assignable to type 'string'
    prop: true,
  },
};

const obj7: {
  prop:
    | {
        type: "foo";
        prop: string;
      }
    | {
        type: "bar";
        prop: number;
      };
} = {
  // current: Type '{ type: "foo"; prop: number; }' is not assignable to type '{ type: "foo"; prop: string; } | { type: "bar"; prop: number; }'.
  //  Types of property 'prop' are incompatible.
  //    Type 'number' is not assignable to type 'string'.
  prop: {
    type: "foo",
    // pr: Type 'number' is not assignable to type 'string'.
    prop: 42,
  },
};
```